data_browser/server.py

```diff
@@ -189,8 +197,30 @@ def read_parquet_file(path: str, offset: int, count: int) -> dict:
     }
 
 
-def has_permissions(path: str, root_paths: list[str]) -> bool:
+def has_permissions(path: str, root_paths: list[str], blocked_paths: list[str] | None = None) -> bool:
     """Returns whether the user can access `path` according to the permissions."""
+
+    # Check if path is blocked
+    if blocked_paths:
+        for blocked_path_pattern in blocked_paths:
+            # For cloud storage paths, check if the path starts with the blocked pattern
+            if path.startswith(CLOUD_STORAGE_PREFIXES):
+                if path.startswith(blocked_path_pattern):
+                    return False
+            else:
+                # For local paths, use os.path.commonpath for consistent logic
+                if not os.path.isabs(path):
+                    # Don't allow relative paths
+                    return False
+                try:
+                    if os.path.commonpath([path, blocked_path_pattern]) == blocked_path_pattern:
+                        # The path is blocked if it's a subpath of the blocked pattern
+                        return False
+                except ValueError:
+                    # Paths don't have a common base, so not blocked by this pattern
+                    continue
```

Comment:
There is code duplication here between blocked and allow...can we factor this out into a separate function?  I think this is a very critical piece of code, so we should invest in making it as simple / easy to understand as possible.